arrow-select/src/concat.rs

```diff
@@ -54,6 +58,53 @@ fn binary_capacity<T: ByteArrayType>(arrays: &[&dyn Array]) -> Capacities {
     Capacities::Binary(item_capacity, Some(bytes_capacity))
 }
 
+fn concat_dictionaries<K: ArrowDictionaryKeyType>(
+    arrays: &[&dyn Array],
+) -> Result<ArrayRef, ArrowError> {
+    let output_len = arrays.iter().map(|x| x.len()).sum();
+    if !should_merge_dictionary_values::<K>(arrays, output_len) {
+        return concat_fallback(arrays, Capacities::Array(output_len));
+    }
+
+    // Recompute dictionaries
+    let dictionaries: Vec<_> = arrays
+        .iter()
+        .map(|a| (as_dictionary_array::<K>(*a), None))
+        .collect();
+
+    let merged = merge_dictionary_values(&dictionaries)?;
+
+    // Recompute keys
+    let mut keys = PrimitiveBuilder::<K>::with_capacity(output_len);
+
+    for ((d, _), mapping) in dictionaries.iter().zip(merged.key_mappings) {
+        for key in d.keys_iter() {
+            keys.append_option(key.map(|x| mapping[x]));
+        }
+    }
```

Comment:
Is there any way to use the new `extend` function that got added? Something like (untested)


```suggestion
    for ((d, _), mapping) in dictionaries.iter().zip(merged.key_mappings) {
        keys.extend(d.keys_iter().map(|x| mapping[x]))
    }
```